### DISCUSSION TITLE
Quick way to remove scratch buffers?

### DISCUSSION THREAD

**Discussion by gonstoll** [folke/snacks.nvim#207] [Q&A] [answered]
The more I use `snacks` the more I like it ❤️ in particular, I really like the `scratch` module! Tho I have a quite cumbersome way to remove scratch buffers 🫤 I just `rm -rf ~/.local/share/nvim/scratch` and remove all, which is not ideal. Ideally it'd be nice to be able to remove them from the list. I use `fzf-lua`, and something I do often is list all the opened buffers and just remove the ones I want to close one by one. In my mind it makes sense something like that for removing scratch buffers, but I don't know how to achieve this. Has anyone been able to accomplish something like that?

**Comment by dpetka2001** [comment, marked as answer]
`snacks` uses `vim.ui.select` to list available Scratch buffers, but as far as I know it doesn't accept actions, like FzfLua or Telescope. 

One solution would be to add a parameter to `M.select(mode)` and then check for example if it is `"delete"` to execute some additional logic to delete the file. But, in my personal opinion, that would soon escalate to other requests about additional actions and maybe that would be out of scope in the end.

Since you use FzfLua, you can take make your own custom picker by using the Scratch API to get the list of items. Something like the following should work I believe
```lua
function M.scratch_delete()
  local entries = {}  -- use this to populate the table with single string, because FzfLua expects items to be strings and not an object like `M.list` returns
  local items = Snacks.scratch.list()
  local item_map = {}  -- use this to store `item`, so we can later get access to `item.file`
  local utils = require("fzf-lua.utils")

  local function hl_validate(hl)
    return not utils.is_hl_cleared(hl) and hl or nil
  end

  local function ansi_from_hl(hl, s)
    return utils.ansi_from_hl(hl_validate(hl), s)
  end

  for _, item in ipairs(items) do
    item.icon = item.icon or Snacks.util.icon(item.ft, "filetype")
    item.branch = item.branch and ("branch:%s"):format(item.branch) or ""
    item.cwd = item.cwd and vim.fn.fnamemodify(item.cwd, ":p:~") or ""
    local display = string.format("%s %s %s %s", item.cwd, item.icon, item.name, item.branch)  -- same as what Snacks.scratch uses to display items
    table.insert(entries, display)
    item_map[display] = item
  end

  local fzf = require("fzf-lua")

  fzf.fzf_exec(entries, {
    prompt = "Scratch Buffers",
    fzf_opts = {
      ["--header"] = string.format(
        ":: <%s> to %s | <%s> to %s",
        ansi_from_hl("FzfLuaHeaderBind", "enter"),
        ansi_from_hl("FzfLuaHeaderText", "Select Scratch"),
        ansi_from_hl("FzfLuaHeaderBind", "ctrl-d"),
        ansi_from_hl("FzfLuaHeaderText", "Delete Scratch")
      ),
    },
    actions = {
      ["default"] = function(selected)
        local item = item_map[selected[1]]
        Snacks.scratch.open({ icon = item.icon, file = item.file, name = item.name, ft = item.ft })
      end,
      ["ctrl-d"] = function(selected)
        local item = item_map[selected[1]]
        os.remove(item.file)
        vim.notify("Deleted scratch file: " .. item.file)
      end,
    },
  })
end
```
And you can add additional actions according to your preferences. I believe the tools are there and each user should create his own custom picker to do what he wants. Bind it to a keymap and it should do what you want
```lua
vim.keymap.set("n", "<leader>D", function()
  Util.fzf.scratch_delete()  -- this is just the path I use to store my custom pickers for fzf, change that to wherever you will store the above function
end, { desc = "Delete scratch file" })
```

**Reply by gonstoll** [reply]
Woww nice ✨ this works perfectly! Thank you so much @dpetka2001 🎉! Any idea how you can keep fzf's picker open after removing? Seems to perform the action and then close automatically. I tried something like

```lua
-- ...
  local actions = require('fzf-lua.actions')

-- ...
      ['ctrl-d'] = function(selected)
        local item = item_map[selected[1]]
        os.remove(item.file)
        vim.notify('Deleted scratch file: ' .. item.file)
        actions.resume()
      end,
```

Seems to stay in the picker, but the list still reflects a stale list.

**Reply by dpetka2001** [reply]
Just add `M.scratch_buffers()` instead of `actions.resume()` and wrap the whole `function(selected)` in curly brackets like `{}`, so the window won't flash. So
```lua
      ["ctrl-d"] = {
        function(selected)
          local item = item_map[selected[1]]
          os.remove(item.file)
          vim.notify("Deleted scratch file: " .. item.file)
          M.scratch_buffers()
        end,
      },
```
I changed the name of the function to `M.scratch_buffers` but you use whatever you've called it.

**Reply by gonstoll** [reply]
I feel the dumbest for not thinking about this 😄 this works perfect, thank you for the help!! ❤️

**Comment by briandipalma** [comment]
This works for me:

```lua
return {
  "folke/snacks.nvim",
  opts = {
    scratch = {
      win = {
        keys = {
          ["delete"] = {
            "<a-x>",
            function(self)
              vim.api.nvim_win_call(self.win, function()
                vim.cmd([[close]])
                os.remove(vim.api.nvim_buf_get_name(self.buf))
              end)
            end,
            desc = "Delete buffer",
            mode = { "n", "x" },
          },
        },
      },
    },
  },
}
```

**Reply by briandipalma** [reply]
Changed to use `"<a-x>"` instead of "X" because of keymap clash.
